In [1]:
from IPython.core.display import HTML
with open ("../style.css", "r") as file:
    css = file.read()
HTML(css)

# General Linear Regression

In this case study we investigate how much of the <em style="color:blue">fuel consumption</em> of a car can be explained by the 

- number of cylinders,
- engine displacement,
- horsepower,
- weight,
- acceleration, and
- the year the car has been introduced into the market.

This data is given in the `CSV` file `cars.csv`.  In this file, the engine displacement is given in *cubic inches* and the weight is given in *pounds*.  The fuel consumption is specified as *miles per gallon* and the acceleration is given as the number of seconds until the car reaches 60 miles per hour.

The module `csv` offers a number of functions for reading and writing <tt>csv</tt> files.

In [2]:
import csv

Below we read the file `cars.csv` and store the fuel consumption in the list `mpg`, while the number of cylinders, the engine displacement,
the power, the weight, the acceleration, and the building year are stored as the list of lists `Features`. Note also that we have added the constant feature $1$ to every list in `Features`.

In [3]:
with open('cars.csv') as input_file:
    reader   = csv.DictReader(input_file, delimiter=',')
    Features = ['cyl', 'displacement', 'hp' ,'weight', 'acc', 'year']
    X        = []
    Y        = []
    for row in reader:
        Y.append(float(row['mpg']))
        X.append([float(row[f]) for f in Features] + [1.0])

In [4]:
X

[[8.0, 307.0, 130.0, 3504.0, 12.0, 70.0, 1.0],
 [8.0, 350.0, 165.0, 3693.0, 11.5, 70.0, 1.0],
 [8.0, 318.0, 150.0, 3436.0, 11.0, 70.0, 1.0],
 [8.0, 304.0, 150.0, 3433.0, 12.0, 70.0, 1.0],
 [8.0, 302.0, 140.0, 3449.0, 10.5, 70.0, 1.0],
 [8.0, 429.0, 198.0, 4341.0, 10.0, 70.0, 1.0],
 [8.0, 454.0, 220.0, 4354.0, 9.0, 70.0, 1.0],
 [8.0, 440.0, 215.0, 4312.0, 8.5, 70.0, 1.0],
 [8.0, 455.0, 225.0, 4425.0, 10.0, 70.0, 1.0],
 [8.0, 390.0, 190.0, 3850.0, 8.5, 70.0, 1.0],
 [8.0, 383.0, 170.0, 3563.0, 10.0, 70.0, 1.0],
 [8.0, 340.0, 160.0, 3609.0, 8.0, 70.0, 1.0],
 [8.0, 400.0, 150.0, 3761.0, 9.5, 70.0, 1.0],
 [8.0, 455.0, 225.0, 3086.0, 10.0, 70.0, 1.0],
 [4.0, 113.0, 95.0, 2372.0, 15.0, 70.0, 1.0],
 [6.0, 198.0, 95.0, 2833.0, 15.5, 70.0, 1.0],
 [6.0, 199.0, 97.0, 2774.0, 15.5, 70.0, 1.0],
 [6.0, 200.0, 85.0, 2587.0, 16.0, 70.0, 1.0],
 [4.0, 97.0, 88.0, 2130.0, 14.5, 70.0, 1.0],
 [4.0, 97.0, 46.0, 1835.0, 20.5, 70.0, 1.0],
 [4.0, 110.0, 87.0, 2672.0, 17.5, 70.0, 1.0],
 [4.0, 107.0, 90.0, 2430.0,

The number of data pairs of the form $\langle \textbf{x}, y \rangle$ that we have read is stored in the variable `m`.

In [5]:
m = len(Y)
m

392

For efficiency reasons we transform the *feature matrix* `X`, which currently is represented as a list of list, into a `NumPy` matrix.

In [6]:
import numpy as np

In [7]:
X = np.array(X)

Note that every row in this matrix contains the data corresponding to a single car.

Since <em style="color:blue">miles per gallon</em> is the inverse of the <em style="color:blue">fuel consumption</em>, the vector `Y` is defined as the reciprocal of the `mpg` values that are currently stored in `Y`. 

In [8]:
Y = np.array([1 / Y[i] for i in range(m)])

The weight vector `w` is specified via the *normal equation*:
$$ (X^\top \cdot X) \cdot \textbf{w} = X^\top \cdot \textbf{y} $$ 
This linear equation can be solved for `w` using the method `np.linalg.solve`.  Note that the *transpose* of the matrix `X` can be computed by writing `X.T`.
Furthermore, *matrix multiplication* of two matrices `A` and `B` is denoted as `A @ B` in `numpy`. 

In [9]:
%%time
w = np.linalg.solve(X.T @ X, X.T @ Y)
print(w)

[ 1.39205897e-03 -1.70283082e-05  1.13756363e-04  1.10909594e-05
  3.38800847e-04 -1.26596723e-03  8.95296095e-02]
CPU times: user 6.73 ms, sys: 1.9 ms, total: 8.63 ms
Wall time: 3.44 ms


The <em style="color:blue">residual sum of squares</em> is given by the following sum:
$$ \texttt{RSS} = \sum\limits_{i=1}^m \Bigl(\bigl(\textbf{x}^{(i)}\bigr)^\top \cdot \textbf{w} - y_i\Bigr)^2 $$
Here $\textbf{x}^{(i)}$ is the $i$-th row of the matrix $X$, while $y_i$ is the $i$-th component of the vector $\textbf{y}$.
The expression $\bigl(\textbf{x}^{(i)}\bigr)^\top \cdot \textbf{w}$ is the predicted value of the linear model, while $y_i$ is the actual value.
As the feature Matrix $X$ is defined as
$$ X = \left( \begin{array}{c}
              \bigl(\textbf{x}^{(1)}\bigr)^\top \\
              \vdots \\
              \bigl(\textbf{x}^{(m)}\bigr)^\top
              \end{array}
       \right)
$$
we can compute the variable `RSS` as follows:

In [10]:
RSS = np.sum((X @ w - Y) ** 2)
RSS

0.012569055391196694

We compute the <em style="color:blue">average fuel consumption</em> $\bar{\mathbf{y}}$ according to the formula:
$$ \bar{\mathbf{y}} = \frac{1}{m} \cdot \sum\limits_{i=1}^m y_i $$ 

In [11]:
yMean = np.mean(Y)
yMean

0.04782242789602457

We  compute the <em style="color:blue">total sum of squares</em> `TSS` according to the following formula:
$$ \mathtt{TSS} := \sum\limits_{i=1}^m \bigl(y_i - \bar{\mathbf{y}}\bigr)^2 $$

In [12]:
TSS = np.sum((Y - yMean) ** 2)
TSS

0.10825652135293372

Now the <em style="color:blue">proportion of explained variance</em> $R^2$ is calculated via the formula:
$$ R^2 = 1 - \frac{\mathtt{RSS}}{\mathtt{TSS}}$$

In [13]:
R2 = 1 - RSS/TSS
R2

0.8838956283269112

It looks like we can explain about $88\%$ of the fuel consumption by the data given in our `CSV` file.  Given that our data does not contain any parameters describing the air resistance we cannot hope to do much better. 